### PR TITLE
Low-hanging Ember optimizations

### PR DESCRIPTION
**Pull request by armanbilge**
Just some micro-optimizations I noticed while working on https://github.com/http4s/http4s/pull/6752.

### CODE REVIEW

**Comment by armanbilge** [comment]
Actually the optimization I really wanted to make is to this:
https://github.com/http4s/http4s/blob/04b2bbbd1787ac7084f589240ed8a062f38ed584/ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala#L78

I was hoping we could go directly from `StringBuilder` to `Array[Byte]` without making an intermediate `String`. One way to do this is to use `CharBuffer.wrap(sb)` with the ISO_8859_1 `CharsetEncoder`. But there is an encoding fast-path for `String` that is not available for a generic `CharBuffer` so I'm skeptical if this would improve performance.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala
Comment (ChristopherDavenport) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:53]: I do not recognize this operator, is this optimized over an if with assignment. If so great, but specialized mutability stuff I don't really know.

Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:53]: Good point. Maybe I should write it like this.
```scala
appliedContentLength = appliedContentLength || h.name == `Content-Length`.name
```

File: ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala
Comment (diesalbla) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala:142]:
```suggestion
  private[this] val lastChunk: Stream[Nothing, Byte] = {
```

**Review by diesalbla — APPROVED** [review]

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala
Comment (armanbilge) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:134]: @danicheg wdyt? I studied the bytecode and looks like these are emitted as private static methods, which I think it's safe to assume that the JIT will inline.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala
Comment (danicheg) [inline ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:131]: Continuing bike-shedding, speaking in terms of saving a few ns, shouldn't we place `!isNoPayloadMethod` ahead of `isEmptyBody`? I mean, `!isNoPayloadMethod` might be `false` a bit frequently then `isEmptyBody`.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala
Comment (danicheg) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:134]: Can confirm – readability is better now :)

File: ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala
Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/Encoder.scala:131]: That might be, but I think reference equality `eq` is faster than set membership `.contains`.

**Comment by armanbilge** [comment]
Apologies, since this PR is sitting I started committing additional optimizations :) each commit should be individual reviewable.

**Review by valencik — APPROVED** [review]
I think this looks good. Lots of `ByteVector.view` to avoid copying arrays and some nice `unchunks`.
It does make me think we could benefit from some more ember benchmarks. I'm going to look into how hard those might be to add.

File: ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala
Comment (armanbilge) [reply ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala]: Thanks yeah, that's a lot nicer 😂

**Comment by rossabaker** [comment]
Anything holding this up from merge?

**Comment by armanbilge** [comment]
Oh, when Andrew first reviewed they weren't a maintainer so I couldn't merge. Now I can!

**Pull request merged**
